php/class-customize-concurrency.php

```diff
@@ -8,10 +8,377 @@
+			<ul>
+				<# _.each( data.notifications, function( notification ) { #>
+					<li class="notice notice-{{ notification.type || 'info' }} {{ data.altNotice ? 'notice-alt' : '' }} notice-concurrency_conflict" data-code="{{ notification.code }}" data-type="{{ notification.type }}">
+						<# if ( /concurrency_conflict/.test( notification.code ) ) { #>
+							<button class="button concurrency-conflict-override" type="button" data-tooltip="Reject Change"><span class="dashicons dashicons-thumbs-down"></span></button>
+							<button class="button concurrency-conflict-accept" type="button" data-tooltip="Accept Change"><span class="dashicons dashicons-thumbs-up"></span></button>
+							Conflict due to concurrent update by {{ notification.data.user }}.
+							<p><b>Change:</b> <i>"{{ notification.data.their_value }}"</i></p>
+						<# } #>
+					</li>
+				<# } ); #>
+			</ul>
+		</script>
+		<?php
+
+	}
+
+	/**
+	 * Prevent sanitize_title() on the post_name.
+	 *
+	 * @see wp_insert_post()
+	 * @filter wp_insert_post_data
+	 *
+	 * @param array $data    An array of slashed post data.
+	 * @param array $postarr An array of sanitized, but otherwise unmodified post data.
+	 * @return array
+	 */
+	public function preserve_inserted_post_name( $data, $postarr ) {
+		if ( self::POST_TYPE === $data['post_type'] && isset( $postarr['post_name'] ) ) {
+			$data['post_name'] = wp_slash( $postarr['post_name'] );
+		}
+		return $data;
+	}
+
+	/**
+	 * Set up conflict checking within snapshots.
+	 *
+	 * @param Customize_Snapshot         $snapshot         Snapshot object.
+	 * @param Customize_Snapshot_Manager $snapshot_manager Snapshot Manager.
+	 */
+	public function customize_snapshot_save_before( Customize_Snapshot $snapshot, Customize_Snapshot_Manager $snapshot_manager ) {
+		$saved_settings = $snapshot->data();
+		$this->sanitize_conflicts( $saved_settings, $snapshot_manager->customize_manager );
+	}
+
+	/**
+	 * Set up conflict checking without snapshots.
+	 *
+	 * @param \WP_Customize_Manager $wp_customize Customizer object.
+	 */
+	public function customize_register( \WP_Customize_Manager $wp_customize ) {
+		$post_values = $wp_customize->unsanitized_post_values();
+		$saved_settings = $this->get_saved_settings( array_keys( $post_values ) );
+		$this->sanitize_conflicts( $saved_settings, $wp_customize );
+	}
+
+
+	/**
+	 * Check all incoming values to see if any were changed between the last time we checked and when we published these
+	 * new values.
+	 *
+	 * @param array                 $saved_settings Array of settings to be checked for conflicts.
+	 * @param \WP_Customize_Manager $wp_customize   Customize object.
+	 */
+	public function sanitize_conflicts( $saved_settings, \WP_Customize_Manager $wp_customize ) {
+		// @todo When publishing a snapshot outside the Customizer (e.g. from admin or WP Cron), the setting $timestamps should get supplied from modified times for setting updates in the snapshot?
```

Comment:
This todo could be referring to two different scenarios.

First, if we are publishing a scenario that does not have any conflicts yet and we just want to store timestamps, we can store the current time as timestamps in our c9y post type so that any live edits that are currently happening will get a notice. I don't see a point in looking back to the point in time when the edit was added to the snapshot. It is going live now and would trigger conflicts with other live changes as of now.

The other scenario involves live (published) changes made in between the time that the snapshot was started and when it is eventually published. This applies to scheduled and admin publishes, but also even to snapshots published in the customizer. When publishing a snapshot, do we:

- a) Look at the time when the snapshot editor last saw a live value (first time they edited that field) and check to see if any changes have gone live since then and throw a conflict when it is published?
- b) Warn them as they are saving to the snapshot that a change has happened on the live side since the snapshot was started?

If b), we need to also consider changes that go live in between scheduling the snapshot and when it eventually goes live. And actually, maybe that is the one point this todo was considering.

If we go with a, I would look back to the first time the field was edited, not the last edit (which is what we are currently storing). If we go with b, we would want to use the last time that the snapshot as a whole was saved since we will be conflict checked up to that time.